cmake/CeleritasUtils.cmake

```
@@ -221,3 +254,66 @@ function(celeritas_configure_file input output)
 endfunction()
 
 #-----------------------------------------------------------------------------#
+
+macro(celeritas_setup_option var option) #[condition]
+  if(${ARGC} EQUAL 2)
+    # always-on-option
+    list(APPEND ${var}_OPTIONS ${option})
+  elseif(${ARGV2})
+    # variable evaluates to true
+    list(APPEND ${var}_OPTIONS ${option})
+  else()
+    list(APPEND ${var}_DISABLED_OPTIONS ${option})
+  endif()
+endmacro()
+
+#-----------------------------------------------------------------------------#
+
+function(celeritas_define_options var doc)
+  if(NOT ${var}_OPTIONS)
+    message(FATAL_ERROR "${var}_OPTIONS has no options")
+  endif()
+  mark_as_advanced(${var}_OPTIONS)
+
+  list(GET ${var}_OPTIONS 0 _default)
+  set(${var} "${_default}" CACHE STRING "${doc}")
+  set_property(CACHE ${var} PROPERTY STRINGS "${${var}_OPTIONS}")
+
+  list(FIND ${var}_OPTIONS "${${var}}" _index)
+  if(_index EQUAL -1)
+    message(SEND_ERROR "Invalid value ${var}=${${var}}: must be "
+      "${${var}_OPTIONS}; overriding for next configure")
+    set(${var} "${_default}" CACHE STRING "${doc}" FORCE)
+  endif()
+endfunction()
+
+#-----------------------------------------------------------------------------#
+
+function(celeritas_generate_option_macros var)
+  # Add disabled options first
+  set(_options)
+  foreach(_val IN LISTS ${var}_DISABLED_OPTIONS)
+    string(TOUPPER "${_val}" _val)
+    list(APPEND _options "#define ${var}_${_val} 0")
```

A more verbose approach (resulting in a more typical use where both `#if not defined(..)` and `#if not ..` work
```suggestion
    list(APPEND _options "#if defined(${var}_${_val})")
    list(APPEND _options "#undef ${var}_${_val} 0")
    list(APPEND _options "#endif")
```

---

Yeah but then you can't use it as a C++ constant... so you won't be able to write `if (CELERITAS_GEO == CELERITAS_GEO_VECGEOM)`.

---

In what context would that make sense (i.e. where the code after the if does not need any vecgeom symbols)?

---

The most common one is for test results that change depending on the selection of the "core" geometry: see https://github.com/celeritas-project/celeritas/pull/726/files#diff-a40e865ce029c1d7d0e6db09a86cb5feb80985417e1b89798194377afc4d0b1e

---

I see.  It does make sense.